### DISCUSSION TITLE
YOLOV8 on openVino runtime Output shape error

### DISCUSSION THREAD

**Discussion by official-MKV** [ultralytics/ultralytics#1228] [Q&A]
I am building a computer-vision detection model. 
I don't have a NVIDIA GPU. I use an intel based graphics card which makes inference on CPU look way better, but still after training a detection model using 'yolov8n.pt'  and running the 'best.pt' model on my cpu, it is still terribly slow. I converted the 'best.pt' model to an openvino IR model to run on openvino-runtime( I read some articles that claim it boosts performance when running on intel-based cpu).
My problem is with the output shape of the IR model. 
```
import cv2
import matplotlib.pyplot as plt
import numpy as np
from openvino.runtime import Core
from ultralytics import YOLO
ie = Core()

model = ie.read_model(model=r"\models\IR\best.xml")
compiled_model = ie.compile_model(model=model, device_name="CPU")

input_layer_ir = compiled_model.input(0)
output_layer_ir = compiled_model.output(0)


image = cv2.imread(r'\Resources\test_image.jpg')
N, C, H, W = input_layer_ir.shape
resized_image = cv2.resize(image, (W, H))
input_image = np.expand_dims(resized_image.transpose(2, 0, 1), 0)
plt.imshow(cv2.cvtColor(image, cv2.COLOR_BGR2RGB));
boxes = compiled_model([input_image])[output_layer_ir]

print(boxes.shape)
```

```
<Output: names[output0] shape[1,6,8400] type: f32>
```
I don't understand the output and how to go about it and it seems to be an output shape particualr to yolov8, other pretrained models don't have this shape. I have tried training using differnt yolov8 models and even tried different datasets but still endup with the same shape that has no meaning to me `[1,6,8400]`.  How do I go about this ? What is this shape

**Comment by derek-knox** [comment]
@official-MKV This issue may help https://github.com/ultralytics/ultralytics/issues/751

As I understand it:
1. `1`: batch (number of inputs where 1 is one image)
    - I'm a little fuzzy on the definition here FYI
2. `6`: box + number of classes (first 4 = xywh of box, last 2 = probability of bounding box against each class idx—0 and 1 respectively)
3. `8400`: number of detections
    - I _believe_ this number is a function of the `stride` value (defaults to 32) and the `input` dimensions

If anyone has more clarity, detail, or correctness to add, please do.

**Reply by deno-source** [reply]
> @official-MKV This issue may help #751
> 
> As I understand it:
> 
> 1. `1`: batch (number of inputs where 1 is one image)
>    
>    * I'm a little fuzzy on the definition here FYI
> 2. `6`: box + number of classes (first 4 = xywh of box, last 2 = probability of bounding box against each class idx—0 and 1 respectively)
> 3. `8400`: number of detections
>    
>    * I _believe_ this number is a function of the `stride` value (defaults to 32) and the `input` dimensions
> 
> If anyone has more clarity, detail, or correctness to add, please do.

Hello, I have also encountered this problem, in tensorflow.js, how can I concretely handle [1,6,8400] and turn them into predictions?

**Comment by derek-knox** [comment]
@official-MKV My updated understanding: `8400` is not the "number of detections" but instead a dimensions size that is a function of the stride. As such this number is a function of the model input dimensions. [By default, the stride is `32`](https://github.com/ultralytics/ultralytics/blob/b8cc404ebffa0b5604528c58729c597cd2b9c6e1/ultralytics/yolo/data/base.py#L35).

So for `stride = 32` we have `640/32` which gives us the base cell size of `20`. Doubling a few times (I'm guessing this step evolved experientially) gives us three total values:
- `20`
- `40`
- `80`

And these are squared and summed to get the final dimension count (`80^2 + 40^2 + 20^2 = 8400`).

So a 256px input results in `256/32 = 8`. So:
- `8`
- `16`
- `32`

And these are squared and multiplied to get the final dimension count (`32^2 + 16^2 + 8^2 = 1344`).

---

![image](https://github.com/ultralytics/ultralytics/assets/21041937/42d26e3c-03dc-4883-8728-3682223f88cb)

https://openmmlab.medium.com/dive-into-yolov8-how-does-this-state-of-the-art-model-work-10f18f74bab1#:~:text=(2)%20Dimensional%20transformation,b%2C8400%2C4)%2C%20respectively.
